pkg/mutation/match/match.go

```diff
@@ -72,6 +72,114 @@ func Matches(match *Match, obj client.Object, ns *corev1.Namespace) (bool, error
 		return false, errors.New("invalid call to Matches(), ns must not be nil for Namespace objects")
 	}
 
+	topLevelMatchers := []matchFunc{
+		kindsMatch,
+		scopeMatch,
+		namespacesMatch,
+		excludedNamespacesMatch,
+		labelSelectorMatch,
+		namespaceSelectorMatch,
+	}
+
+	for _, fn := range topLevelMatchers {
+		ok, err := fn(match, obj, ns)
+		if err != nil {
+			return false, err
+		}
+		if !ok {
+			return false, nil
+		}
+	}
+
+	return true, nil
+}
+
+// matchFunc defines the matching logic of a Top Level Matcher.  A TLM receives the match criteria,
+// an object, and the namespace of the object and decides if there is a reason why the object does
+// not match.  If the TLM associated with the matching function is not defined by the user, the
+// matchFunc should return true.
+type matchFunc func(match *Match, obj client.Object, ns *corev1.Namespace) (bool, error)
+
+func namespaceSelectorMatch(match *Match, obj client.Object, ns *corev1.Namespace) (bool, error) {
+	if match.NamespaceSelector == nil {
+		return true, nil
+	}
+
+	clusterScoped := ns == nil || isNamespace(obj)
```

Comment:
Are there any cases where an object is cluster scoped but `ns != nil`? Could we just have

```
clusterScoped := ns == nil
```
?

Reply:
The [`mutateRequest` function in `pkg/webhook/mutation.go`](https://github.com/open-policy-agent/gatekeeper/blob/master/pkg/webhook/mutation.go#L147) seems to hold the answer to this.

It looks like the `ns` variable _is_ set when the admission request satisfies `req.Kind.Kind == namespaceKind && req.Kind.Group == ""`.

So, while more intuitive, I think your suggested change would incorrectly handle namespace objects.